### PR TITLE
Барсуков М.Н. Лаб. 2 Группа 6313

### PR DESCRIPTION
**Pull request by Mishele123**
**ФИО:** Барсуков Михаил Николаевич
**Номер группы:** 6313
**Номер лабораторной:** 2
**Краткое описание исходного кода:** Я взял исходный код из *своей старой лабораторной работы*. Вот ссылка на него: https://github.com/Mishele123/python_lab5
**Краткое описание проделанной работы:** В ходе работы над улучшением предоставленного кода были выполнены следующие действия: оптимизация кода, исправления нарушения DRY и SOLID, добавление обработки исключений и др.

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import os
+import glob
+from PIL import Image
+import random
+
+
+class dataset(torch.utils.data.Dataset):
```
Comment (Mishele123) [inline lab_2/main.py:19]: нарушение наименование класса

`class Dataset(torch.utils.data.Dataset):`

File: lab_2/main.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    def __init__(self,file_list,transform=None):
+        self.file_list = file_list
+        self.transform = transform
+        
+    def __len__(self):
+        self.filelength = len(self.file_list)
+        return self.filelength
+    
+    def __getitem__(self,idx):
```
Comment (Mishele123) [inline lab_2/main.py:28]: Проблема: Отсутствие обработки ошибок
Если изображение не может быть загружено, код завершится с ошибкой. 
Решение: Добавить блок try-except.
```

def __getitem__(self, idx):
    img_path = self.file_list[idx]
    try:
        img = Image.open(img_path)
    except Exception as ex:
        print(f"Error loading image {img_path}: {ex}")
        return None, None
```

File: lab_2/main.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+import pandas as pd
+import matplotlib.pyplot as plt
+import os
+import glob
+from PIL import Image
+import random
+
+
+class dataset(torch.utils.data.Dataset):
+    def __init__(self,file_list,transform=None):
```
Comment (Mishele123) [inline lab_2/main.py:20]: неясное название переменной file_list.
Название file_list не отражает, что это список путей к изображениям.
Необходимо переименовать переменную для ясности.

`def __init__(self, image_paths, transform=None):`

File: lab_2/main.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    def __getitem__(self,idx):
+        img_path = self.file_list[idx]
+        img = Image.open(img_path)
+        img_transformed = self.transform(img)
+        
+        label = img_path.split("\\")[1].split("_")[0]
+        if label == 'leopard':
+            label=1
+        elif label == 'tiger':
+            label=0
```
Comment (Mishele123) [inline lab_2/main.py:37]: Нарушение DRY: Логика присвоения метки дублируется
необходимо использовать словарь для сопоставления меток```

label_map = {'leopard': 1, 'tiger': 0}
    label = label_map.get(img_path.split("\\")[1].split("_")[0], -1) 
```
# -1 для нераспознанных меток

File: lab_2/main.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+    test_transforms = transforms.Compose([   
+        transforms.Resize((240, 240)),
+        transforms.RandomResizedCrop(240),
+        transforms.RandomHorizontalFlip(),
+        transforms.ToTensor()
+        ])
+    
+    
+    return train_transforms, validate_transforms, test_transforms
```
Comment (Mishele123) [inline lab_2/main.py:127]: Устранение дублирования кода
создать отдельную функцию для преобразований данных, чтобы избежать дублирования
```
def get_transforms() -> transforms.Compose:
    return transforms.Compose([
        transforms.Resize((240, 240)),
        transforms.RandomResizedCrop(240),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
    ])

train_transforms = get_transforms()
validate_transforms = get_transforms()
test_transforms = get_transforms()
```

Comment (Mishele123) [inline lab_2/main.py:127]: Устранить дублирование кода
Создать отдельную функцию для преобразования данных для избежания дублирования

```
def get_transforms() -> transforms.Compose:
    return transforms.Compose([
        transforms.Resize((240, 240)),
        transforms.RandomResizedCrop(240),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
    ])

train_transforms = get_transforms()
validate_transforms = get_transforms()
test_transforms = get_transforms()
```

File: lab_2/main.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    "the function returns an array containing path to the images (split datas)"
+    from sklearn.model_selection import train_test_split
+
+    train_dir = "train_dir" # path to dir with images
+
+    train_list = glob.glob(os.path.join(train_dir,'*.jpg'))
+    train_list, temp_list = train_test_split(train_list, test_size=0.2)
+    test_list, validate_list = train_test_split(temp_list, test_size=0.5)
+
+    # print(train_list[0].split("\\")[1].split("_")[0])
```
Comment (Mishele123) [inline lab_2/main.py:141]: Удаление мертвого кода.
Удалить во всем файле закомментированные строки и ненужные комментарии, чтобы сделать код чище.

File: lab_2/main.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    train_list, test_list, validate_list = return_lists()
+    train_transforms, test_transforms, validate_transforms = data_augmentation()
+
+    train_data = dataset(train_list, transform=train_transforms)
+    test_data = dataset(test_list, transform=test_transforms)
+    validate_data = dataset(validate_list, transform=validate_transforms)
+
+    train_loader = torch.utils.data.DataLoader(dataset = train_data, batch_size=batch_size, shuffle=True )
+    test_loader = torch.utils.data.DataLoader(dataset = test_data, batch_size=batch_size, shuffle=True)
+    validate_loader = torch.utils.data.DataLoader(dataset = validate_data, batch_size=batch_size, shuffle=True)
```
Comment (Mishele123) [inline lab_2/main.py:97]: Проблема: длинные строки кода (не удовлетворяющие pep8)
Решение: Разбить длинные строки на несколько более коротких. (и так по всему файлу)

File: lab_2/main.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        ])
+    
+    
+    return train_transforms, validate_transforms, test_transforms
+
+
+def return_lists() -> list:
+    
+    "the function returns an array containing path to the images (split datas)"
+    from sklearn.model_selection import train_test_split
```
Comment (Mishele123) [inline lab_2/main.py:133]: Проблема: Импорт библиотек внутри функции может снизить производительность, если функция вызывается многократно.
Решение: Переместить импорт в начало файла.

File: lab_2/main.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        transforms.RandomResizedCrop(240),
+        transforms.RandomHorizontalFlip(),
+        transforms.ToTensor()
+        ])
+    
+    
+    return train_transforms, validate_transforms, test_transforms
+
+
+def return_lists() -> list:
```
Comment (Mishele123) [inline lab_2/main.py:130]: Проблема: Функция возвращает три списка, но тип возвращаемого значения указан как list.
Решение: Изменить аннотацию типа на tuple
`def return_lists() -> tuple:`

File: lab_2/main.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+    
+    "the function returns an array containing path to the images (split datas)"
+    from sklearn.model_selection import train_test_split
+
+    train_dir = "train_dir" # path to dir with images
+
+    train_list = glob.glob(os.path.join(train_dir,'*.jpg'))
+    train_list, temp_list = train_test_split(train_list, test_size=0.2)
+    test_list, validate_list = train_test_split(temp_list, test_size=0.5)
```
Comment (Mishele123) [inline lab_2/main.py:139]: Проблема: Отсутствие обработки ошибок может привести к сбоям, если директория не существует или не содержит изображений.
Решение: Добавить блок try-except, для отлавливания ошибок
```
try:
        train_list = glob.glob(os.path.join(train_dir, '*.jpg'))
        
        if not train_list:
            raise FileNotFoundError(f"No images found in directory: {train_dir}")

        train_list, temp_list = train_test_split(train_list, test_size=0.2)
        test_list, validate_list = train_test_split(temp_list, test_size=0.5)

    except Exception as ex:
        print(f"An error occurred: {ex}")
        return [], [], []```

File: lab_2/main.py [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+    print("Предсказанные вероятности:", probabilities)
+    predicted_class = torch.argmax(probabilities).item()
+    if predicted_class == 1:
+        print("Предсказанный класс: leopard")
+    else:
+        print("Предсказанный класс: tiger")
+
+
+def draw_graph(train_values : [], val_values : [], title : str, xlabel : str, ylabel : str) -> None:
```
Comment (Mishele123) [inline lab_2/main.py:300]: Использование пустых списков:
В аннотациях типов вместо [] следует использовать List из модуля typing. Использование [] не является корректным типом.

`def draw_graph(train_values: List[float], val_values: List[float], title: str, xlabel: str, ylabel: str) -> None:`

File: lab_2/main.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+            output = model(data)
+            loss = criterion(output, label)
+            
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            
+            acc = ((output.argmax(dim=1) == label).float().mean())
+            epoch_accuracy += acc/len(train_loader)
+            epoch_loss += loss/len(train_loader)
```
Comment (Mishele123) [inline lab_2/main.py:178]: Проблема: Повторные вычисления len(train_loader) в цикле обучения.
Вычислите его один раз перед циклом.

```
num_batches = len(train_loader)
for epoch in range(epochs):
    ...
    for data, label in train_loader:
        ...
```

File: lab_2/main.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+def image_test_probs(leopard_probs: []) -> None:
+    test_list = return_lists()[1]
+
+    idx = list(map(lambda x: x[0],leopard_probs))
+    prob = list(map(lambda x: x[1],leopard_probs))
+
+    submission = pd.DataFrame({'id':idx,'label':prob})
+    
+    test_list = return_lists()[1]
```
Comment (Mishele123) [inline lab_2/main.py:234]: повторный вызов функции return_lists

File: lab_2/main.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+        for data, fileid in test_loader:
+            data = data.to(device)
+            preds = model(data)
+            preds_list = F.softmax(preds, dim=1)[:, 1].tolist()
+            leopard_probs += list(zip(list(fileid), preds_list))
+    leopard_probs.sort(key = lambda x : int(x[0]))
+
+    image_test_probs(leopard_probs)
+
+    torch.save(model.state_dict(), 'trained_model.pth')
```
Comment (Mishele123) [inline lab_2/main.py:223]: функция не удовлетворяет принципам solid
Разбить код на отдельные функции

`def main() -> None:
    device = "cpu"
    lr = 0.001
    batch_size = 100
    epochs = 10

    train_loader, test_loader, validate_loader = create_data_loaders(batch_size)
    model = initialize_model(device)
    optimizer, criterion = initialize_optimizer(model, lr)

    train_and_validate_model(model, train_loader, validate_loader, optimizer,
                                             criterion, epochs, device)
    evaluate_model(model, test_loader, device)
    save_model(model)`

File: lab_2/main.py [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+
+def image_test_probs(leopard_probs: []) -> None:
```
Comment (Mishele123) [inline lab_2/main.py:226]: использовать List из библиотеки typing

File: lab_2/main.py [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+        print(f"Selected index: {i}")
+
+        label = submission.loc[submission.index == i, 'label'].values[0]
+        
+        print(f"Label = {label}")
+  
+        if label > 0.5:
+            label = 1
+        else:
+            label = 0
```
Comment (Mishele123) [inline lab_2/main.py:252]: Проблема: Проверка метки (if label > 0.5) может быть упрощена.
Исправление: Преобразовать значение в 0 или 1 можно с помощью int().

File: lab_2/main.py [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+            
+        img_path = os.path.join(test_list[i])
+        print(f"Image path: {img_path}")
+        
+        img = Image.open(img_path)
+        ax.set_title(class_[label])
+        ax.imshow(img)
```
Comment (Mishele123) [inline lab_2/main.py:259]: нет обработки ошибок если изображение не найдено
Добавить обработку исключений:

```
try:
    img = Image.open(img_path)
    ax.imshow(img)
    ax.set_title(class_map[label])
except FileNotFoundError:
    print(f"Файл не найден: {img_path}")
    ax.set_title("Изображение не найдено")
    ax.imshow(np.zeros((240, 240, 3)))
```

File: lab_2/main.py [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+    image_path = "D:\\python_labs\\datas\\leopard\\leopard_1039.jpg"
+
+
+    # Определение преобразований для входного изображения
+    transform = transforms.Compose([
+        transforms.Resize((240, 240)),
+        transforms.RandomResizedCrop(240),
+        transforms.RandomHorizontalFlip(),
+        transforms.ToTensor(),
+    ])
```
Comment (Mishele123) [inline lab_2/main.py:280]: дублирование кода
лучше вызвать функцию get_transforms()

File: lab_2/main.py [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+        ax.imshow(img)
+
+    plt.show()
+
+
+def image_recognition() -> None:
+    "Test model"
+    model = Cnn().to('cpu')
+
+    model.load_state_dict(torch.load('trained_model.pth'))
```
Comment (Mishele123) [inline lab_2/main.py:268]: добавить обработку ошибок при загрузке модели

**Review by IlyaOv — APPROVED** [review]
Зачтено
